storage/src/mmr/hasher.rs

```diff
@@ -1,82 +1,213 @@
 //! Decorator for a cryptographic hasher that implements the MMR-specific hashing logic.
 
+use crate::mmr::{
+    iterator::{leaf_num_to_pos, leaf_pos_to_num},
+    Error, Storage,
+};
 use commonware_cryptography::Hasher as CHasher;
+use std::future::Future;
 
-/// Hasher decorator the MMR uses for computing leaf, node and root hashes.
-pub(crate) struct Hasher<'a, H: CHasher> {
+/// A trait for computing the various digests of an MMR.
+pub trait Hasher<H: CHasher>: Send + Sync {
+    /// Computes the digest for a leaf given its position and the element it represents.
+    fn leaf_digest(
+        &mut self,
+        pos: u64,
+        element: &[u8],
+    ) -> impl Future<Output = Result<H::Digest, Error>> + Send;
+
+    /// Computes the digest for a node given its position and the digests of its children.
+    fn node_digest(&mut self, pos: u64, left: &H::Digest, right: &H::Digest) -> H::Digest;
+
+    /// Computes the root for an MMR given its size and an iterator over the digests of its peaks in
+    /// decreasing order of height.
+    fn root_digest<'b>(
+        &mut self,
+        pos: u64,
+        peak_digests: impl Iterator<Item = &'b H::Digest>,
+    ) -> H::Digest;
+
+    /// Compute the digest of a byte slice.
+    fn digest(&mut self, data: &[u8]) -> H::Digest;
+
+    /// Access the underlying commonware_cryptography hasher.
+    fn c_hasher(&mut self) -> &mut H;
```

Comment:
nit: recommend calling this `inner()`

Reply:
I "think" this is only used in tests but am ok exporting it (in case it is needed for some reason later).

Reply:
done

Reply:
yes it's only used in testing ATM, felt it's likely to be useful more generally though.